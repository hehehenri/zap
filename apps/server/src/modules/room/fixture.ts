import { RoomDefinition, RoomModel } from "./RoomModel";

type Fields = Pick<RoomDefinition, Exclude<keyof RoomDefinition, "_id">>

export const createRoom = async (fields?: DeepPartial<Fields>) => {
  return await new RoomModel(fields).save();
}
