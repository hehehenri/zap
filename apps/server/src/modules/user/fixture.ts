import config from "../../config";
import { UserDefinition, UserModel } from "./UserModel"
import bcrypt from "bcryptjs";

let counter = 0;

const getCounter = () => {
  counter += 1;

  return counter;
}

type Fields = Pick<UserDefinition, Exclude<keyof UserDefinition, "_id">>

export const createUser = async (fields?: DeepPartial<Fields>) => {
  const index = getCounter();

  const plainTextPassword = fields?.password ?? "password";

  const password = bcrypt.hashSync(plainTextPassword, config.jwt.saltRounds);
  const username = fields?.username ?? `user#${index}`;

  const user = new UserModel({ password, username });
  await user.save();

  return user;
}
