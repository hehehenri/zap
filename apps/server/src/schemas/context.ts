import { Context } from "koa";
import WebSocket, { WebSocketServer } from "ws";

export type GraphQLContext = {
  user: any | null,
}

export type KoaContext = {
  websocket?: {
    server: WebSocketServer;
    ws: () => Promise<WebSocket>;
  }  
} & Context;
