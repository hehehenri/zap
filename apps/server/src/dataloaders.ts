import { UserLoader } from "./modules/user/UserLoader"

export type DataLoaders = {
  UserLoader: ReturnType<typeof UserLoader.getLoader>
}

export const buildLoaders = (): DataLoaders => {
  return {
    UserLoader: UserLoader.getLoader(),
  }
}
