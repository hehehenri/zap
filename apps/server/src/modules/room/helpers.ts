import { UserDocument } from "../user/UserModel";
import { RoomDocument } from "./RoomModel";

export const isRoomMember = (room: RoomDocument, user: UserDocument) => {
  return room
    .participants
    .map(u => u._id.toString())
    .includes(user._id.toString());
}
