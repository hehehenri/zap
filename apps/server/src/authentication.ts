import jwt from "jsonwebtoken";
import { UserModel, UserDocument } from "./modules/user/UserModel";
import config from "./config";
import { KoaContext } from "./context";

export type Config = {
  secret: string
}

type TokenPayload = {
  id: string
}

export const getToken = (ctx: KoaContext) => {
  const auth = ctx.headers.authorization;
  if (!auth) return null;

  if (auth.startsWith("Bearer ")) {
    return auth.substring(7, auth.length);
  }

  return null
}

export const getAuth = async (token: string | null) => {
  if (!token) return { user: null };

  try {
    const decoded = jwt.verify(token, config.jwt.secret) as TokenPayload;

    const user = await UserModel.findById(decoded.id);
    return { user };
  } catch (_) {
    return { user: null };
  }
}

export const generateToken = (user: UserDocument) => {
  return jwt.sign({ id: user._id }, config.jwt.secret);
}
