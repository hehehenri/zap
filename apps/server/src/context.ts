import { Context as KoaDefaultContext } from "koa";
import WebSocket, { WebSocketServer } from "ws";
import { UserDocument } from "./modules/user/UserModel";
import { getAuth, getToken } from "./authentication";
import { DataLoaders, buildLoaders } from "./dataloaders";

export type KoaContext = {
  websocket?: {
    server: WebSocketServer;
    ws: () => Promise<WebSocket>;
  }  
} & KoaDefaultContext ;

export type Context = {
  user: UserDocument | null,
  dataloaders: DataLoaders
}

export const initialContext = () => {
  return {
    user: null,
    dataloaders: buildLoaders(),
  }
}

export const buildContext = async (ctx: KoaContext): Promise<Context> => {
  const context = initialContext();
  
  const token = getToken(ctx);  
  const { user } = await getAuth(token, context);

  return {
    ...context,
    user,
  };
}
