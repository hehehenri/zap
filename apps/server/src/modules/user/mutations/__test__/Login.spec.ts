import { createUser } from "../../fixture";
import { describeWithDb, testQuery } from "../../../../test/helpers";
import { getAuth } from "../../../../authentication";
import { buildContext, initialContext } from "@/context";
import { buildLoaders } from "@/dataloaders";

const query = ({ username, password }: { username?: string, password?: string }) => ({
  query: `
    mutation Login($username: String! $password: String!) {
      login(input: { username: $username, password: $password}) { token }
    }
  `,
  variables: {
    username: username ?? "username",
    password: password ?? "password",
  }
})

describeWithDb('user/mutations/login', () => {
  const context = initialContext();
  
  it("should return user token", async () => {
    const data = { username: "username", password: "password"};
    const user = await createUser(data);
      
    const response = await testQuery({ query: query(data) });

    expect(response.status).toBe(200);

    const token = response.body.data.login.token;
    const { user: u } = await getAuth(token, context); 

    expect(u?.username).toBe(user.username);
  });

  it(
    "should return invalid payload when username or password is invalid",
    async () => {
      const data = { username: "username", password: "password" };
      await createUser(data);
      
      const response = await testQuery({ query: query({ ...data, password: "invalidPassword" }) })

      expect(response.status).toBe(422);
      expect(response.body.errors).toMatchSnapshot();
    });
  }
);
