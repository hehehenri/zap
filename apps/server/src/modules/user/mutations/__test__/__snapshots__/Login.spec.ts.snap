// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user/mutations/login should return invalid payload when username or password is invalid 1`] = `
[
  {
    "locations": [
      {
        "column": 7,
        "line": 3,
      },
    ],
    "message": "Invalid password or user not found",
    "path": [
      "login",
    ],
  },
]
`;
